hazelcast/src/test/java/com/hazelcast/partition/PartitionMigrationListenerTest.java

```diff
@@ -330,6 +331,76 @@ public void testRemoveMigrationListener() {
         verify(listener, never()).replicaMigrationCompleted(any(ReplicaMigrationEvent.class));
     }
 
+    /**
+     * @see <a href="https://hazelcast.atlassian.net/browse/HZ-2651">HZ-2651 - MigrationListener: Difference between
+     *      "wall clock elapsed time" and the totalElasedTime API</a>
+     */
+    @Test
+    public void testMigrationListenerElapsedTime() throws InterruptedException, ExecutionException, TimeoutException {
+        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();
+
+        // Use an arbitrarily high number of partitions to exacerbate the issue
+        final Config config = new Config().setProperty(ClusterProperty.PARTITION_COUNT.getName(), String.valueOf(1000));
+
+        final HazelcastInstance hz1 = createPausedMigrationCluster(factory, Optional.of(config));
+
+        final Stopwatch migrationTimer = Stopwatch.createUnstarted();
+        final CompletableFuture<MigrationState> migrationStateReference = new CompletableFuture<>();
+
+        hz1.getPartitionService().addMigrationListener(new MigrationListener() {
+            @Override
+            public void migrationStarted(final MigrationState migrationState) {
+                // Don't trust the listener start time because no guarantees as to when its executed
+            }
+
+            @Override
+            public void migrationFinished(final MigrationState migrationState) {
+                migrationTimer.stop();
+                migrationStateReference.complete(migrationState);
+            }
+
+            @Override
+            public void replicaMigrationCompleted(final ReplicaMigrationEvent replicaMigrationEvent) {
+            }
+
+            @Override
+            public void replicaMigrationFailed(final ReplicaMigrationEvent replicaMigrationEvent) {
+            }
+        });
+
+        LOGGER.fine("Starting second instance...");
+        final HazelcastInstance hz2 = factory.newHazelcastInstance(config);
+
+        // Back to ACTIVE
+        migrationTimer.start();
+        changeClusterStateEventually(hz2, ClusterState.ACTIVE);
+
+        LOGGER.fine("Awaiting migration completion...");
+        final Duration reportedMigrationTime = Duration
+                .ofMillis(migrationStateReference.get(ASSERT_TRUE_EVENTUALLY_TIMEOUT, TimeUnit.SECONDS).getTotalElapsedTime());
+
+        assertFalse(reportedMigrationTime.isZero());
+
+        final String message = MessageFormat.format("migrationState.getTotalElapsedTime={1}, migrationTimer={0}",
+                migrationTimer.elapsed(), reportedMigrationTime);
+
+        LOGGER.fine(message);
+        assertTrue(message, migrationTimer.elapsed().compareTo(reportedMigrationTime) >= 0);
```

Comment:
The assertion message here should be more relevant - something like `Migration timer ({0}) was less than reported elapsed time ({1})` (also make sure the output in the message is human-friendly, not `Duration#toString()` format)

Reply:
> The assertion message here should be more relevant - something like Migration timer ({0}) was less than reported elapsed time ({1})

Good idea - done.

> also make sure the output in the message is human-friendly, not Duration#toString() format

`Duration#toString()` is in ISO8601 format, so it's not too bad (e.g. `PT8H6M12.345S`). I can use `org.apache.commons.lang3.time.DurationFormatUtils#formatDurationHMS` so it's in something like `00:00:00.204` but I'm not sure that's really any better, especially in a test?

Reply:
It's not the worst, but it's not the easiest to read either 😅 I think just using `#toMillis()` and presenting as Milliseconds is fine; generally we present test timings in plain values like that anyway.

Reply:
`migrationState.getTotalElapsedTime=190 milliseconds, migrationTimer=779 milliseconds`